osutil/mkdir_test.go

```diff
@@ -0,0 +1,258 @@
+	c.Assert(osutil.IsDir(tmpDir+"/foo/bar"), check.Equals, true)
+
+	err = osutil.Mkdir(tmpDir+"/foo/bar", 0o755, nil)
+	c.Assert(err, check.ErrorMatches, `.*: file exists`)
+}
+
+func (mkdirSuite) TestParentExistsButNotDir(c *check.C) {
+	tmpDir := c.MkDir()
+
+	_, err := os.Create(tmpDir + "/foo")
+	c.Assert(err, check.IsNil)
+
+	err = osutil.Mkdir(tmpDir+"/foo/bar/", 0o755, &osutil.MkdirOptions{
+		MakeParents: true,
+	})
+	c.Assert(err, check.ErrorMatches, `.*: not a directory`)
+}
+
+func (mkdirSuite) TestChmod(c *check.C) {
+	oldmask := syscall.Umask(0022)
+	defer syscall.Umask(oldmask)
+
+	tmpDir := c.MkDir()
+
+	err := osutil.Mkdir(tmpDir+"/foo", 0o777, &osutil.MkdirOptions{
+		Chmod: true,
+	})
+	c.Assert(err, check.IsNil)
+	c.Assert(osutil.IsDir(tmpDir+"/foo"), check.Equals, true)
+
+	info, err := os.Stat(tmpDir + "/foo")
+	c.Assert(err, check.IsNil)
+	c.Assert(info.Mode().Perm(), check.Equals, os.FileMode(0o777))
+}
+
+func (mkdirSuite) TestNoChmod(c *check.C) {
+	oldmask := syscall.Umask(0022)
+	defer syscall.Umask(oldmask)
+
+	tmpDir := c.MkDir()
+
+	err := osutil.Mkdir(tmpDir+"/foo", 0o777, nil)
+	c.Assert(err, check.IsNil)
+	c.Assert(osutil.IsDir(tmpDir+"/foo"), check.Equals, true)
+
+	info, err := os.Stat(tmpDir + "/foo")
+	c.Assert(err, check.IsNil)
+	c.Assert(info.Mode().Perm(), check.Equals, os.FileMode(0o755))
+}
+
+func (mkdirSuite) TestMakeParentsAndChmod(c *check.C) {
+	tmpDir := c.MkDir()
+
+	err := osutil.Mkdir(tmpDir+"/foo/bar", 0o777, &osutil.MkdirOptions{
+		MakeParents: true,
+		Chmod:       true,
+	})
+	c.Assert(err, check.IsNil)
+	c.Assert(osutil.IsDir(tmpDir+"/foo"), check.Equals, true)
+	c.Assert(osutil.IsDir(tmpDir+"/foo/bar"), check.Equals, true)
+
+	info, err := os.Stat(tmpDir + "/foo")
+	c.Assert(err, check.IsNil)
+	c.Assert(info.Mode().Perm(), check.Equals, os.FileMode(0o777))
+
+	info, err = os.Stat(tmpDir + "/foo/bar")
+	c.Assert(err, check.IsNil)
+	c.Assert(info.Mode().Perm(), check.Equals, os.FileMode(0o777))
+}
+
+func (mkdirSuite) TestMakeParentsAndNoChmod(c *check.C) {
+	oldmask := syscall.Umask(0022)
+	defer syscall.Umask(oldmask)
+
+	tmpDir := c.MkDir()
+
+	err := osutil.Mkdir(tmpDir+"/foo/bar", 0o777, &osutil.MkdirOptions{
+		MakeParents: true,
+	})
+	c.Assert(err, check.IsNil)
+	c.Assert(osutil.IsDir(tmpDir+"/foo"), check.Equals, true)
+	c.Assert(osutil.IsDir(tmpDir+"/foo/bar"), check.Equals, true)
+
+	info, err := os.Stat(tmpDir + "/foo")
+	c.Assert(err, check.IsNil)
+	c.Assert(info.Mode().Perm(), check.Equals, os.FileMode(0o755))
+
+	info, err = os.Stat(tmpDir + "/foo/bar")
+	c.Assert(err, check.IsNil)
+	c.Assert(info.Mode().Perm(), check.Equals, os.FileMode(0o755))
+}
+
+func (mkdirSuite) TestMakeParentsChmodAndChown(c *check.C) {
+	if os.Getuid() != 0 {
+		c.Skip("requires running as root")
+	}
+
+	username := os.Getenv("PEBBLE_TEST_USER")
+	group := os.Getenv("PEBBLE_TEST_GROUP")
+	if username == "" || group == "" {
+		c.Fatalf("must set PEBBLE_TEST_USER and PEBBLE_TEST_GROUP")
+	}
```

Comment:
note to reviewers: this test only passes currently because we don't run tests as root. In a follow-up PR I would like to change this to `XGO_*` and setup root tests in our CI.

Reply:
I do like the plain `TEST_USER` and `TEST_GROUP` variables. Though when testing the other projects I expect this test doesn't run, right?

Reply:
I'd be happy with `TEST_USER` and `TEST_GROUP` (but I don't think there's any problem with `XGO_*`, as these wouldn't be used in Pebble or snapd if they're tested here).

That reminds me, this is a "root test" -- we should probably also port that GitHub job over from Pebble so these tests are actually run in CI. [See the workflow config.](https://github.com/canonical/pebble/blob/39bce1b4f234c9c52be919caccb818fcc3b9a59e/.github/workflows/tests.yml#L27-L41)